sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveFetchCursor.scala

```diff
@@ -0,0 +1,101 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.SparkException
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.expressions.VariableReference
+import org.apache.spark.sql.catalyst.plans.logical.{FetchCursor, LogicalPlan, SingleStatement}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.COMMAND
+import org.apache.spark.sql.connector.catalog.CatalogManager
+import org.apache.spark.sql.errors.QueryCompilationErrors.unresolvedVariableError
+
+/**
+ * Resolves the target SQL variables in FetchCursor command.
+ */
+class ResolveFetchCursor(val catalogManager: CatalogManager) extends Rule[LogicalPlan]
+  with ColumnResolutionHelper {
+  private val variableResolution = new VariableResolution(catalogManager.tempVariableManager)
+
+  /**
+   * Checks for duplicate variable names and throws an exception if found.
+   * Names are normalized when the variables are created.
+   * No need for case insensitive comparison here.
+   */
+  private def checkForDuplicateVariables(variables: Seq[VariableReference]): Unit = {
+    val dups = variables.groupBy(_.identifier).filter(kv => kv._2.length > 1)
+    if (dups.nonEmpty) {
+      throw new AnalysisException(
+        errorClass = "DUPLICATE_ASSIGNMENTS",
+        messageParameters = Map("nameList" ->
+          dups.keys.map(key => toSQLId(key.name())).mkString(", ")))
+    }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(COMMAND), ruleId) {
+    // Resolve FetchCursor wrapped in SingleStatement
+    case s @ SingleStatement(fetchCursor: FetchCursor)
+        if !fetchCursor.targetVariables.forall(_.resolved) =>
+      val resolvedVars = fetchCursor.targetVariables.map {
+        case u: UnresolvedAttribute =>
+          variableResolution.lookupVariable(
+            nameParts = u.nameParts
+          ) match {
+            case Some(variable) => variable.copy(canFold = false)
+            case _ => throw unresolvedVariableError(u.nameParts, Seq("SYSTEM", "SESSION"))
+          }
+
+        case other => throw SparkException.internalError(
+          "Unexpected target variable expression in FetchCursor: " + other)
+      }
+
+      s.copy(parsedPlan = fetchCursor.copy(targetVariables = resolvedVars))
+
+    // Check for duplicates after resolution
+    case s @ SingleStatement(fetchCursor: FetchCursor)
+        if fetchCursor.targetVariables.forall(_.resolved) =>
+      val targetVariables = fetchCursor.targetVariables.map(_.asInstanceOf[VariableReference])
+      checkForDuplicateVariables(targetVariables)
```

Comment:
can we do it right after we resolve the variables? why adding another case match?